src/core/mapping/instance_manager.cc

```diff
@@ -72,29 +89,49 @@ struct construct_overlapping_region_group_fn {
     const InstanceSet::Domain& domain,
     const std::map<InstanceSet::RegionGroupP, InstanceSet::InstanceSpec>& instances)
   {
-    auto bound = domain.bounds<DIM, coord_t>();
-    std::vector<InstanceSet::Region> regions(1, region);
+    auto bound       = domain.bounds<DIM, coord_t>();
+    size_t bound_vol = bound.volume();
+    std::set<InstanceSet::Region> regions{region};
+
+#ifdef DEBUG_LEGATE
+    log_instmgr.debug() << "construct_overlapping_region_group( " << region << "," << domain << ")";
+#endif
 
     for (const auto& pair : instances) {
       auto& group = pair.first;
 
-      Rect<DIM> group_bbox = group->bounding_box;
-      auto intersect       = bound.intersection(group_bbox);
-      if (intersect.empty()) continue;
-
-      // Don't merge if the unused space would be more than the space saved
+      Rect<DIM> group_bbox = group->bounding_box.bounds<DIM, coord_t>();
+#ifdef DEBUG_LEGATE
+      log_instmgr.debug() << "  check intersection with " << group_bbox;
+#endif
+      auto intersect = bound.intersection(group_bbox);
+      if (intersect.empty()) {
+#ifdef DEBUG_LEGATE
+        log_instmgr.debug() << "    no intersection";
+#endif
+        continue;
+      }
+
+      // Only allow merging if the bloating isn't "too big"
       auto union_bbox  = bound.union_bbox(group_bbox);
       size_t bound_vol = bound.volume();
       size_t union_vol = union_bbox.volume();
-
-      // If it didn't get any bigger then we can keep going
-      if (bound_vol == union_vol) continue;
```

Comment:
@manopapad this heuristic is removed to handle the case where an instance was originally created for a subregion and later another instance is requested for a bigger subregion.